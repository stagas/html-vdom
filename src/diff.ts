import { VList } from './v'

// import { inspect } from 'util'
// const log = (...args: any[]) => console.log(...args.map(x => inspect(x, false, Infinity, true)))

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ParentLike = ParentNode | VList<any>

const $ = (P: ParentLike) => (Array.isArray(P) ? P.at(-1) : P.lastChild)

// const nextSibling = <T>(parent: ParentLike<T>, child: T | ChildNode): T | ChildNode | null =>
//   parent instanceof VList ? parent.nextSiblingOf(child as T) : (child as ChildNode).nextSibling

const _ = <T>(parent: ParentLike | T[], index: number): T | null =>
  Array.isArray(parent)
    ? (parent[index] as T) ?? null
    : ((parent as ParentNode).childNodes[index] as unknown as T) ?? null

const len = (P: ParentLike) => (Array.isArray(P) ? P.length : P.childNodes.length)

export const diff = <T>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  eq: (a: any, b: any) => boolean,
  P: ParentLike,
  b: T[],
  a: T[] = []
) => {
  // log(a, b)

  // A holds the anchor element where insertBefore(x, A) latches on
  let A: T | null

  // B holds the current element in the new list b
  let B: T | null

  // C holds a temp value of the list a that will be compared and maybe inserted
  let C: T | null

  // log(b, a)
  let i = 0

  const n = b.length
  let p = a.length

  //              p
  // a  1 2 3 4 5
  //        i
  // b  1 2 0 4 5
  //              n
  // R  1 2
  //
  // advance start pointer while equal head
  // TODO:                     set A B here
  while (i < n && i < p && eq((A = _(a, i)), (B = _(b, i)))) i++

  //          p
  // a  1 2 3 4 5
  //        i
  // b  1 2 0 4 5
  //          n
  // R  1 2
  //
  // shorten end pointer while equal tail
  // while (i < n && i < p && eq(_(a, p - 1), _(b, n - 1))) --p, --n

  // insert:
  //        A
  //        p
  // a  1 2 4 5
  //              i
  // b  1 2 0 0 0 4 5
  //              n
  // R  1 2 0 0 0
  //

  // append:
  //          A
  //          p
  // a  1 2 3
  //          i
  // b  1 2 3 4
  //            n
  //

  // prepend:
  //
  //    A
  //    p
  // a  1 2 3
  //    i
  // b  4 1 2 3
  //      n
  //

  // insert/append/prepend shortcut
  if (i == p) {
    A = _(a, p)
    for (; i < n; i++) P.insertBefore(_(b, i)!, A)
    // we are done -- rest of elements (if any) are the same
  } else {
    // replace:
    //
    //      A
    //        p
    // a  1 2 3
    //      i
    // b  1 4 3
    //        n
    //      B

    // replace + remaining tail:
    //
    //        A
    //        p
    // a  1 2 3
    //        i
    // b  1 4 5 6 3
    //            n
    // R  1 4 3
    //      B

    //      x
    //      A
    //          p
    // a  1 2 3 4
    //        i
    //        C
    //            y
    // b  1 3 2 4
    //          n
    //            B
    // R  1 3 2 4

    //      x
    //      A
    //            p
    // a  1 2 3 4
    //        C
    //        i
    //      y
    // b  1 3 4 2
    //      B
    //            n
    // R  1 3 4 2

    // moves:
    //
    //        x
    //                    p
    //        A
    // a  1 2 3 4 5 6 7 8
    //    i         y
    //    C
    // b  3 6 5 1 2 7 8 4
    //              B
    //                    n
    //
    // R  1 2 3 4 5 6 7 8
    // R  3 6 5 1 2 4 7 8
    //

    // let x = i
    // let y = i
    // for (; y < n; y++) {
    //   for (i = y; i < p; i++) {
    //     C = _(P, i)
    //     if (eq(C, B)) {
    //       P.insertBefore(C, A)
    //       break
    //     }
    //     //
    //   }

    //   while (y < n && x < p && eq((A = _(P, y)), (B = _(b, y)))) x++, y++

    //   //      A = _(P, x)
    // }

    let x = 0
    out: while (((A = _(P, i)!), (B = _(b, i)!), i++ < n)) {
      if (!eq(A, B)) {
        for (x = i; x < p; x++) {
          C = _(P, x)!
          if (eq(B, C)) {
            P.insertBefore(C, A)
            continue out
          }
        }
        for (x = i; x < n; x++) {
          C = _(b, x)
          if (eq(A, C)) {
            P.insertBefore(B, A)
            continue out
          }
        }
        if (A) P.replaceChild(B, A)
        else P.appendChild(B) //, null)
      }
    }
    p = len(P)
    i = n
    while (i++ < p) P.removeChild($(P))

    // remove tail:
    //
    //          p
    // a  1 2 3
    //      i
    // b  1 2
    //        n

    // remove head:
    //
    //      p
    // a  1 2 3
    //    i
    // b  2 3
    //    n
    // for (; i < p--; ) P.removeChild(_(P, i))
  }

  // if (P instanceof Node) {
  //   log('RESULT', P.innerHTML)
  // } else log('RESULT', [...P])
}
