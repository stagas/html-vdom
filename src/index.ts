export * from './h'
export * from './render'
