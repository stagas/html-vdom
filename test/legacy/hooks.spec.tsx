/* eslint-disable @typescript-eslint/no-explicit-any */
import { Fragment, current, h, render, useCallback, useHook } from '../../src'

let c: any
beforeEach(() => (c = document.createElement('div')))

const clean = (s: string) =>
  s.replaceAll('<!--[-->', '').replaceAll('<!--]-->', '').replaceAll('<!---->', '')

describe('hooks', () => {
  it('p', () => {
    let hook: any
    let content = 'foo'
    let called = 0
    const Foo = () => {
      hook = current.hook
      called++
      return <p>{content}</p>
    }
    render(<Foo />, c)
    expect(called).toEqual(1)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    content = 'bar'
    hook.trigger()
    expect(called).toEqual(2)
    expect(clean(c.innerHTML)).toMatchSnapshot()
  })

  it('same tree', () => {
    const hooks: any[] = []
    let content = 'foo'
    let calledFoo = 0
    let calledBar = 0
    const Bar = () => {
      if (!hooks.includes(current.hook)) hooks.push(current.hook)
      calledBar++
      return <p>{content}</p>
    }
    const Foo = () => {
      if (!hooks.includes(current.hook)) hooks.push(current.hook)
      calledFoo++
      return <Bar />
    }
    render(<Foo />, c)
    expect(calledFoo).toEqual(1)
    expect(calledBar).toEqual(1)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    expect(hooks.length).toEqual(2)
    content = 'bar'
    hooks.splice(0).forEach(hook => hook.trigger())
    expect(hooks.length).toEqual(2)
    expect(calledFoo).toEqual(2)
    expect(calledBar).toEqual(3)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    expect(hooks.length).toEqual(2)
    expect(calledFoo).toEqual(2)
    expect(calledBar).toEqual(3)
    content = 'zoo'
    hooks.splice(0).forEach(hook => hook.trigger())
    expect(hooks.length).toEqual(2)
    expect(calledFoo).toEqual(3)
    expect(calledBar).toEqual(5)
    expect(clean(c.innerHTML)).toMatchSnapshot()
  })

  it('123 different tree', () => {
    const hooks: any[] = []
    let content = 'foo'
    let calledFoo = 0
    let calledBar = 0
    const Bar = () => {
      if (!hooks.includes(current.hook)) hooks.push(current.hook)
      calledBar++
      return <p>{content}</p>
    }
    const Foo = () => {
      if (!hooks.includes(current.hook)) hooks.push(current.hook)
      calledFoo++
      return (
        <div>
          <Bar />
        </div>
      )
    }
    render(<Foo />, c)
    expect(calledFoo).toEqual(1)
    expect(calledBar).toEqual(1)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    expect(hooks.length).toEqual(2)
    content = 'bar'
    hooks.splice(0).forEach(hook => hook.trigger())
    expect(hooks.length).toEqual(2)
    expect(calledFoo).toEqual(2)
    expect(calledBar).toEqual(3)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    expect(calledFoo).toEqual(2)
    expect(calledBar).toEqual(3)
    content = 'zoo'
    hooks.splice(0).forEach(hook => hook.trigger())
    expect(hooks.length).toEqual(2)
    expect(calledFoo).toEqual(3)
    expect(calledBar).toEqual(5)
    expect(clean(c.innerHTML)).toMatchSnapshot()
  })

  it('same tree using useHook', () => {
    const hooks: any[] = []
    let content = 'foo'
    let calledFoo = 0
    let calledBar = 0
    const Bar = () => {
      const hook = useHook()
      if (!hooks.includes(hook)) hooks.push(hook)
      calledBar++
      return <p>{content}</p>
    }
    const Foo = () => {
      const hook = useHook()
      if (!hooks.includes(hook)) hooks.push(hook)
      calledFoo++
      return <Bar />
    }
    render(<Foo />, c)
    expect(calledFoo).toEqual(1)
    expect(calledBar).toEqual(1)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    expect(hooks.length).toEqual(2)
    content = 'bar'
    hooks.splice(0).forEach(hook => hook())
    expect(hooks.length).toEqual(2)
    expect(calledFoo).toEqual(2)
    expect(calledBar).toEqual(3)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    expect(hooks.length).toEqual(2)
    expect(calledFoo).toEqual(2)
    expect(calledBar).toEqual(3)
    content = 'zoo'
    hooks.splice(0).forEach(hook => hook())
    expect(hooks.length).toEqual(2)
    expect(calledFoo).toEqual(3)
    expect(calledBar).toEqual(5)
    expect(clean(c.innerHTML)).toMatchSnapshot()
  })

  it('different tree using useHook', () => {
    const hooks: any[] = []
    let content = 'foo'
    let calledFoo = 0
    let calledBar = 0
    const Bar = () => {
      const hook = useHook()
      if (!hooks.includes(hook)) hooks.push(hook)
      calledBar++
      return <p>{content}</p>
    }
    const Foo = () => {
      const hook = useHook()
      if (!hooks.includes(hook)) hooks.push(hook)
      calledFoo++
      return (
        <div>
          <Bar />
        </div>
      )
    }
    render(<Foo />, c)
    expect(calledFoo).toEqual(1)
    expect(calledBar).toEqual(1)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    expect(hooks.length).toEqual(2)
    content = 'bar'
    hooks.splice(0).forEach(hook => hook())
    expect(hooks.length).toEqual(2)
    expect(calledFoo).toEqual(2)
    expect(calledBar).toEqual(3)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    expect(hooks.length).toEqual(2)
    expect(calledFoo).toEqual(2)
    expect(calledBar).toEqual(3)
    content = 'zoo'
    hooks.splice(0).forEach(hook => hook())
    expect(hooks.length).toEqual(2)
    expect(calledFoo).toEqual(3)
    expect(calledBar).toEqual(5)
    expect(clean(c.innerHTML)).toMatchSnapshot()
  })

  it('p inc a number', () => {
    let hook: any
    let i = 0
    const Foo = () => {
      hook = current.hook
      return <p>{i++}</p>
    }
    render(<Foo />, c)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook.trigger()
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook.trigger()
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook.trigger()
    expect(clean(c.innerHTML)).toMatchSnapshot()
  })

  it('fragment top', () => {
    let hook: any
    let i = 0
    const Foo = () => {
      hook = current.hook
      return <>{i++}</>
    }
    render(<Foo />, c)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook.trigger()
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook.trigger()
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook.trigger()
    expect(clean(c.innerHTML)).toMatchSnapshot()
  })

  it('fragment deep level 1', () => {
    let hook: any
    let i = 0
    const Foo = () => {
      hook = current.hook
      return <>{i++}</>
    }
    render(
      <div>
        <Foo />
      </div>,
      c
    )
    expect(clean(c.innerHTML)).toEqual('<div>0</div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div>1</div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div>2</div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div>3</div>')
  })

  it('fragment deep level 2', () => {
    let hook: any
    let i = 0
    const Foo = () => {
      hook = current.hook
      return <>{i++}</>
    }
    render(
      <div>
        <p>
          <Foo />
        </p>
      </div>,
      c
    )
    expect(clean(c.innerHTML)).toEqual('<div><p>0</p></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><p>1</p></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><p>2</p></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><p>3</p></div>')
  })

  it('fragment deep level 3', () => {
    let hook: any
    let i = 0
    const Foo = () => {
      hook = current.hook
      return <>{i++}</>
    }
    render(
      <div>
        <p>
          <b>
            <Foo />
          </b>
        </p>
      </div>,
      c
    )
    expect(clean(c.innerHTML)).toEqual('<div><p><b>0</b></p></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><p><b>1</b></p></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><p><b>2</b></p></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><p><b>3</b></p></div>')
  })

  it('fragment siblings (last)', () => {
    let hook: any
    let i = 0
    const Foo = () => {
      hook = current.hook
      return <>{i++}</>
    }
    render(
      <div>
        {[1, 2, 3].map(x => (
          <li key={x}>
            <Foo />
          </li>
        ))}
      </div>,
      c
    )
    expect(clean(c.innerHTML)).toEqual('<div><li>0</li><li>1</li><li>2</li></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><li>0</li><li>1</li><li>3</li></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><li>0</li><li>1</li><li>4</li></div>')
  })

  it('fragment siblings (first)', () => {
    let hook: any
    let i = 0
    const Foo = ({ id }: any) => {
      if (id === 1) hook = current.hook
      return <>{i++}</>
    }
    render(
      <div>
        {[1, 2, 3].map(x => (
          <li key={x}>
            <Foo id={x} />
          </li>
        ))}
      </div>,
      c
    )
    expect(clean(c.innerHTML)).toEqual('<div><li>0</li><li>1</li><li>2</li></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><li>3</li><li>1</li><li>2</li></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><li>4</li><li>1</li><li>2</li></div>')
  })

  it('fragment siblings (middle)', () => {
    let hook: any
    let i = 0
    const Foo = ({ id }: any) => {
      if (id === 2) hook = current.hook
      return <>{i++}</>
    }
    render(
      <div>
        {[1, 2, 3].map(x => (
          <li key={x}>
            <Foo id={x} />
          </li>
        ))}
      </div>,
      c
    )
    expect(clean(c.innerHTML)).toEqual('<div><li>0</li><li>1</li><li>2</li></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><li>0</li><li>3</li><li>2</li></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><li>0</li><li>4</li><li>2</li></div>')
  })

  it('fragment siblings (top last)', () => {
    let hook: any
    let i = 0
    const Foo = () => {
      hook = current.hook
      return <>{i++}</>
    }
    render(
      <div>
        {[1, 2, 3].map(x => (
          <Foo key={x} />
        ))}
      </div>,
      c
    )
    expect(clean(c.innerHTML)).toEqual('<div>012</div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div>013</div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div>014</div>')
  })

  it('fragment siblings (top first)', () => {
    let hook: any
    let i = 0
    const Foo = ({ id }: any) => {
      if (id === 1) hook = current.hook
      return <>{i++}</>
    }
    render(
      <div>
        {[1, 2, 3].map(x => (
          <Foo key={x} id={x} />
        ))}
      </div>,
      c
    )
    expect(clean(c.innerHTML)).toEqual('<div>012</div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div>312</div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div>412</div>')
  })

  it('fragment siblings (top middle)', () => {
    let hook: any
    let i = 0
    const Foo = ({ id }: any) => {
      if (id === 2) hook = current.hook
      return <>{i++}</>
    }
    render(
      <div>
        {[1, 2, 3].map(x => (
          <Foo key={x} id={x} />
        ))}
      </div>,
      c
    )
    expect(clean(c.innerHTML)).toEqual('<div>012</div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div>032</div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div>042</div>')
  })

  it('multiple deep', () => {
    let hook_bar: any
    let hook_foo: any
    let i = 0
    const Bar = () => {
      hook_bar = hook_bar ?? current.hook
      return <b>{i++}</b>
    }
    const Foo = () => {
      hook_foo = hook_foo ?? current.hook
      return (
        <p>
          {i++}
          <Bar />
        </p>
      )
    }
    render(<Foo />, c)
    expect(clean(c.innerHTML)).toEqual('<p>0<b>1</b></p>')
    hook_bar.trigger()
    expect(clean(c.innerHTML)).toEqual('<p>0<b>2</b></p>')

    hook_foo.trigger()
    expect(clean(c.innerHTML)).toEqual('<p>3<b>4</b></p>')

    hook_bar.trigger()
    expect(clean(c.innerHTML)).toEqual('<p>3<b>5</b></p>')
  })

  it('deep and change', () => {
    let hook_bar: any
    let hook_foo: any
    let i = 0
    const Bar = () => {
      hook_bar = current.hook
      return <b>{i++}</b>
    }
    const Foo = () => {
      hook_foo = current.hook
      return <p>{i++}</p>
    }
    render(<Foo />, c)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook_foo.trigger()
    expect(clean(c.innerHTML)).toMatchSnapshot()
    render(<Bar />, c)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook_bar.trigger()
    expect(clean(c.innerHTML)).toMatchSnapshot()
  })

  it('fragment multiple', () => {
    let hook: any
    let i = 0
    const Foo = () => {
      hook = current.hook
      return (
        <>
          {i++}
          {i++}
        </>
      )
    }
    render(<Foo />, c)
    expect(clean(c.innerHTML)).toEqual('01')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('23')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('45')
  })

  it('fragment multiple deep level 1', () => {
    let hook: any
    let i = 0
    const Foo = () => {
      hook = current.hook
      return (
        <>
          {i++}
          {i++}
        </>
      )
    }
    render(
      <div>
        <Foo />
      </div>,
      c
    )
    expect(clean(c.innerHTML)).toEqual('<div>01</div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div>23</div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div>45</div>')
  })

  it('fragment multiple deep level 2', () => {
    let hook: any
    let i = 0
    const Foo = () => {
      hook = current.hook
      return (
        <>
          {i++}
          {i++}
        </>
      )
    }
    render(
      <div>
        <p>
          <Foo />
        </p>
      </div>,
      c
    )
    expect(clean(c.innerHTML)).toEqual('<div><p>01</p></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><p>23</p></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><p>45</p></div>')
  })

  it('fragment siblings multiple deep level 2', () => {
    let hook: any
    let i = 0
    const Foo = () => {
      hook = current.hook
      return (
        <>
          {i++}
          {i++}
        </>
      )
    }
    render(
      <div>
        <p>
          {[1, 2, 3].map(x => (
            <Foo key={x} />
          ))}
        </p>
      </div>,
      c
    )
    expect(clean(c.innerHTML)).toEqual('<div><p>012345</p></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><p>012367</p></div>')
  })

  it('fragment siblings multiple deep level 2 under element', () => {
    let hook: any
    let i = 0
    const Foo = () => {
      hook = current.hook
      return (
        <>
          {i++}
          {i++}
        </>
      )
    }
    render(
      <div>
        <p>
          {[1, 2, 3].map(x => (
            <li key={x}>
              <Foo key={x} />
            </li>
          ))}
        </p>
      </div>,
      c
    )
    expect(clean(c.innerHTML)).toEqual('<div><p><li>01</li><li>23</li><li>45</li></p></div>')
    hook.trigger()
    expect(clean(c.innerHTML)).toEqual('<div><p><li>01</li><li>23</li><li>67</li></p></div>')
  })

  it('useHook', () => {
    let hook: any
    let i = 0
    const Foo = () => {
      hook = useHook()
      return <p>{i++}</p>
    }
    render(<Foo />, c)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook()
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook()
    expect(clean(c.innerHTML)).toMatchSnapshot()
  })

  it('useCallback', () => {
    let fn: any
    let i = 0
    const Foo = () => {
      fn = useCallback(() => i++)
      return <p>{i}</p>
    }
    render(<Foo />, c)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    fn()
    expect(clean(c.innerHTML)).toMatchSnapshot()
    fn()
    expect(clean(c.innerHTML)).toMatchSnapshot()
  })

  it('useCallback passes arguments and this', () => {
    let fn: any
    let i = 0
    let target
    let self
    const Foo = () => {
      fn = useCallback(function (this: any, e: any) {
        i++
        target = e.target
        self = this
      })
      return <button onclick={fn}>{i}</button>
    }
    render(<Foo />, c)
    const button = c.querySelector('button')
    expect(button.nodeName).toEqual('BUTTON')
    expect(button.textContent).toEqual('0')
    expect(clean(c.innerHTML)).toMatchSnapshot()
    button.click()
    expect(button.textContent).toEqual('1')
    expect(target).toBe(button)
    expect(self).toBe(button)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    button.click()
    expect(clean(c.innerHTML)).toMatchSnapshot()
    expect(target).toBe(button)
    expect(self).toBe(button)
  })

  it('useCallback return value on original handler (sanity)', () => {
    let fn: any
    let i = 0
    let target
    let self
    let prevented = false
    const onClick = (e: any) => {
      prevented = e.defaultPrevented
    }
    const Foo = () => {
      fn = useCallback(function (this: any, e: any) {
        i++
        target = e.target
        self = this
      })
      return (
        <div onclick={onClick}>
          <button onclick={fn}>{i}</button>
        </div>
      )
    }
    render(<Foo />, c)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    const button = c.querySelector('button')
    expect(button.nodeName).toEqual('BUTTON')
    expect(button.textContent).toEqual('0')
    button.click()
    expect(button.textContent).toEqual('1')
    expect(target).toBe(button)
    expect(self).toBe(button)
    expect(prevented).toBe(false)
  })

  it('useCallback return value on original handler (return false)', () => {
    let fn: any
    let i = 0
    let target
    let self
    let prevented = false
    const onClick = (e: any) => {
      prevented = e.defaultPrevented
    }
    const Foo = () => {
      fn = useCallback(function (this: any, e: any) {
        i++
        target = e.target
        self = this
        return false
      })
      return (
        <div onclick={onClick}>
          <button onclick={fn}>{i}</button>
        </div>
      )
    }
    render(<Foo />, c)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    const button = c.querySelector('button')
    expect(button.nodeName).toEqual('BUTTON')
    expect(button.textContent).toEqual('0')
    button.click()
    expect(button.textContent).toEqual('1')
    expect(target).toBe(button)
    expect(self).toBe(button)
    expect(prevented).toBe(true)
  })

  it('deep hook 2 layers', () => {
    let hook: any
    let i = 0
    const Bar = () => {
      hook = useHook()
      return <p>{i++}</p>
    }
    const Foo = () => {
      return <Bar />
    }
    render(<Foo />, c)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook()
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook()
    expect(clean(c.innerHTML)).toMatchSnapshot()
  })

  it('deep hook 3 layers', () => {
    let hook: any
    let i = 0
    const Zoo = () => {
      hook = useHook()
      return <p>{i++}</p>
    }
    const Bar = () => {
      return <Zoo />
    }
    const Foo = () => {
      return <Bar />
    }
    render(<Foo />, c)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook()
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook()
    expect(clean(c.innerHTML)).toMatchSnapshot()
  })

  it('deep hook 3 layers, top div', () => {
    let hook: any
    let i = 0
    const Zoo = () => {
      hook = useHook()
      return <p>{i++}</p>
    }
    const Bar = () => {
      return <Zoo />
    }
    const Foo = () => {
      return (
        <div>
          <Bar />
        </div>
      )
    }
    render(<Foo />, c)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook()
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook()
    expect(clean(c.innerHTML)).toMatchSnapshot()
  })

  it('prevent retrigger', () => {
    let hook: any
    let i = 0
    const Foo = () => {
      if (!hook) hook = useHook()
      return <p>{i++}</p>
    }
    render(<Foo />, c)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook()
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hook()
    expect(clean(c.innerHTML)).toMatchSnapshot()
  })

  it('empty fragment that updates', () => {
    let hook: any
    let cond = false
    const Foo = () => {
      hook = current.hook
      return !cond ? <></> : <div>ok</div>
    }
    render(<Foo />, c)
    expect(c.textContent).toEqual('')
    cond = true
    hook.trigger()
    expect(c.textContent).toEqual('ok')
  })

  it('fragment that updates deep', () => {
    let hook: any
    let cond = false
    const Bar = () => {
      hook = current.hook
      return !cond ? <></> : <div>ok</div>
    }
    const Foo = () => {
      return <Bar />
    }
    render(<Foo />, c)
    expect(c.textContent).toEqual('')
    cond = true
    hook.trigger()
    expect(c.textContent).toEqual('ok')
  })

  it('xx fragment that updates deep fragment', () => {
    let hook: any
    let cond = false
    const called: string[] = []
    const Bar = () => {
      hook = current.hook
      called.push('bar')
      return !cond ? <></> : <>ok</>
    }
    const Foo = () => {
      called.push('foo')
      return <Bar />
    }
    render(<Foo />, c)
    expect(c.textContent).toEqual('')
    expect(called).toEqual(['foo', 'bar'])

    cond = true
    hook.trigger()
    expect(c.textContent).toEqual('ok')
    expect(called).toEqual(['foo', 'bar', 'bar'])
  })

  it('fragment that updates deep fragment, first div', () => {
    let hook: any
    let cond = false
    const Bar = () => {
      hook = current.hook
      return !cond ? <div></div> : <>ok</>
    }
    const Foo = () => {
      return <Bar />
    }
    render(<Foo />, c)
    expect(c.textContent).toEqual('')
    cond = true
    hook.trigger()
    expect(c.textContent).toEqual('ok')
  })

  it('multiple hooks same parent', () => {
    let hookA: any
    let hookB: any
    let i = 0
    const Foo = () => {
      hookA = useHook()
      hookB = useHook()
      return <p>{i++}</p>
    }
    render(<Foo />, c)
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hookA()
    expect(clean(c.innerHTML)).toMatchSnapshot()
    hookB()
    expect(clean(c.innerHTML)).toMatchSnapshot()
  })
})
