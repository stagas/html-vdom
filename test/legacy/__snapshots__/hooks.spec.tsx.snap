// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hooks 123 different tree 1`] = `"<div><p>foo</p></div>"`;

exports[`hooks 123 different tree 2`] = `"<div><p>bar</p></div>"`;

exports[`hooks 123 different tree 3`] = `"<div><p>zoo</p></div>"`;

exports[`hooks deep and change 1`] = `"<p>0</p>"`;

exports[`hooks deep and change 2`] = `"<p>1</p>"`;

exports[`hooks deep and change 3`] = `"<b>2</b>"`;

exports[`hooks deep and change 4`] = `"<b>3</b>"`;

exports[`hooks deep hook 2 layers 1`] = `"<p>0</p>"`;

exports[`hooks deep hook 2 layers 2`] = `"<p>1</p>"`;

exports[`hooks deep hook 2 layers 3`] = `"<p>2</p>"`;

exports[`hooks deep hook 3 layers 1`] = `"<p>0</p>"`;

exports[`hooks deep hook 3 layers 2`] = `"<p>1</p>"`;

exports[`hooks deep hook 3 layers 3`] = `"<p>2</p>"`;

exports[`hooks deep hook 3 layers, top div 1`] = `"<div><p>0</p></div>"`;

exports[`hooks deep hook 3 layers, top div 2`] = `"<div><p>1</p></div>"`;

exports[`hooks deep hook 3 layers, top div 3`] = `"<div><p>2</p></div>"`;

exports[`hooks different tree using useHook 1`] = `"<div><p>foo</p></div>"`;

exports[`hooks different tree using useHook 2`] = `"<div><p>bar</p></div>"`;

exports[`hooks different tree using useHook 3`] = `"<div><p>zoo</p></div>"`;

exports[`hooks fragment top 1`] = `"0"`;

exports[`hooks fragment top 2`] = `"1"`;

exports[`hooks fragment top 3`] = `"2"`;

exports[`hooks fragment top 4`] = `"3"`;

exports[`hooks multiple hooks same parent 1`] = `"<p>0</p>"`;

exports[`hooks multiple hooks same parent 2`] = `"<p>1</p>"`;

exports[`hooks multiple hooks same parent 3`] = `"<p>2</p>"`;

exports[`hooks p 1`] = `"<p>foo</p>"`;

exports[`hooks p 2`] = `"<p>bar</p>"`;

exports[`hooks p inc a number 1`] = `"<p>0</p>"`;

exports[`hooks p inc a number 2`] = `"<p>1</p>"`;

exports[`hooks p inc a number 3`] = `"<p>2</p>"`;

exports[`hooks p inc a number 4`] = `"<p>3</p>"`;

exports[`hooks prevent retrigger 1`] = `"<p>0</p>"`;

exports[`hooks prevent retrigger 2`] = `"<p>1</p>"`;

exports[`hooks prevent retrigger 3`] = `"<p>2</p>"`;

exports[`hooks same tree 1`] = `"<p>foo</p>"`;

exports[`hooks same tree 2`] = `"<p>bar</p>"`;

exports[`hooks same tree 3`] = `"<p>zoo</p>"`;

exports[`hooks same tree using useHook 1`] = `"<p>foo</p>"`;

exports[`hooks same tree using useHook 2`] = `"<p>bar</p>"`;

exports[`hooks same tree using useHook 3`] = `"<p>zoo</p>"`;

exports[`hooks useCallback 1`] = `"<p>0</p>"`;

exports[`hooks useCallback 2`] = `"<p>1</p>"`;

exports[`hooks useCallback 3`] = `"<p>2</p>"`;

exports[`hooks useCallback passes arguments and this 1`] = `"<button onclick=\\"\\">0</button>"`;

exports[`hooks useCallback passes arguments and this 2`] = `"<button onclick=\\"\\">1</button>"`;

exports[`hooks useCallback passes arguments and this 3`] = `"<button onclick=\\"\\">2</button>"`;

exports[`hooks useCallback return value on original handler (return false) 1`] = `"<div onclick=\\"\\"><button onclick=\\"\\">0</button></div>"`;

exports[`hooks useCallback return value on original handler (sanity) 1`] = `"<div onclick=\\"\\"><button onclick=\\"\\">0</button></div>"`;

exports[`hooks useHook 1`] = `"<p>0</p>"`;

exports[`hooks useHook 2`] = `"<p>1</p>"`;

exports[`hooks useHook 3`] = `"<p>2</p>"`;
